### PR TITLE
Adds Body Marking Customization, Head Accessories, Revises Unathi Customization, Related Minor Fixes, etc.

### PR DESCRIPTION
**Pull request by KasparoVy**
This PR was originally a rough translation of TG-style Lizard People customization to Bay-style Unathi customization :crocodile:. The features were inspired by TG Lizard People customization and more were added to suit our wants/needs.

**Adds the following:**
* The ability to give flagged species markings and head accessories.
* Species-specific body markings.
* Moves Vulpkanin and Tajara earfluff to head accessories.
* Duplicates all other Vulpkanin facial markings into head accessories to make combining two marking patterns a thing.
* Adds frills/spines for Unathi on the facial hair layer, and horns on the head accessory layer.
* Splits facial hair updates into their own proc, and adjusts proc calls accordingly. This was done in order to have horns overlap hair, but be overlapped by facial hair so frills would not be covered by head accessories.
* Small fixes to accommodate PR (#2910) that went through while this was being made so as to not undo or break their work.
* Fixes that will ensure proper ID card icon generation with head accessories and markings, everything on their correct layer.
* Cloning support so players with markings and head accessories get them with their new bodies.
* Accommodations for PR 3025.
* An update to the genetics wiki page to cover all the UI edits, which will be deployed after the PR merge.
* *This PR, however, does not fix the issue with tails not being generated in ID card icons.*

'*' >> Unathi tail customization has been gutted and saved for a later PR because it'll take a while to colour-correct all the sprites (animated ones, too) and work out a system by which to implement tail markings so they use the same colour as body markings while still being animated by body_accessories, etc.

**WORK STILL TO DO:**
- ~~Compatibility with the SQL~~
- ~~Gut tails for now~~
- ~~Compatibility with PRs awaiting merge~~
- ~~Resolve conflicts~~
- ~~Possibly accommodate #3020~~
- ~~Accommodate #3025~~
- ~~Get markings and head accessories to carry through cloning~~
- ~~Update from master~~
- ~~Resolve last conflicts~~
- ~~Final tests~~
- ~~Implement last requested feature~~
- ~~Write update to Wiki Genetics Guide~~

**DONE** :checkered_flag:

**PROOF OF CONCEPT:**

*Character Generation -- (out of date, most recent version features the ability to colour horns individually)*
lizzercustomizationpoc2

*IN-GAME 1, with HORNS, HAIR, FRILLS, AND TIGER-STYLE BODY MARKINGS*
lizzercustomizationpoc4

*IN-GAME 2, with HORNS, HAIR, FRILLS, AND TIGER-STYLE BODY MARKINGS*
lizzercustomizationpoc1

*ID-CARD ICON GENERATION*
The icon is of poor quality and the body colour is oddly bright.
lizzercustomizationpoc3

The arrows indicate spots where the body markings and other newly implemented features can be seen. Markings are barely visible due to the poor quality of the icon.
lizzercustomizationpoc5

-> **See below posts for more example photos** <-

### CODE REVIEW

**Comment by KasparoVy** [comment]
To fix conflicts should I just update branch with everything but #2910?

Additionally, might have to accommodate #3020 as well

**Comment by Crazylemon64** [comment]
I think your best bet might be to move your work in files that conflict to a backup file, pull down the newest edition of the file, and copy over your new icons into it.

File: code/modules/client/preferences_savefile.dm [hunk 1]
```diff
@@ -134,6 +137,8 @@
 	S["skin_red"]			>> r_skin
 	S["skin_green"]			>> g_skin
 	S["skin_blue"]			>> b_skin
+	S["horn_style_name"]	>> horns
+	S["marking_style_name"]	>> m_style
```
Comment (tigercat2000) [inline code/modules/client/preferences_savefile.dm]: don't really need to bother with the savefile at all, it's thoroughly ripped out already and no one else bothered to update it

Comment (Crazylemon64) [inline code/modules/client/preferences_savefile.dm]: Might as well rip it out while I'm at it in my PR

**Comment by tigercat2000** [comment]
@KasparoVy You should try to just update your branch normally, and then run `git status` to figure out what's conflicting.

File: code/modules/client/preferences_savefile.dm (hunk 1 above)
Comment (KasparoVy) [inline code/modules/client/preferences_savefile.dm]: So I'd be alright to just straight delete this file?

**Comment by KasparoVy** [comment]
Updated everything but #2910 that I accommodated

File: code/modules/client/preferences_savefile.dm (hunk 1 above)
Comment (Crazylemon64) [inline code/modules/client/preferences_savefile.dm]: Yeah, go for it

**Comment by KasparoVy** [comment]
Everything is now working as intended

Hair below horns, horns below facial hair, everything gets its own colour
lizzercustomizationpoc6

Just need to get it working with SQL now

**Comment by Fox-McCloud** [comment]
I do wonder if the horns should be renamed to something else and opened up to other species.

Reason I say this is that there's some things that fit poorly as "facial hair", but are none the less lumped in there because nothing else works.

Prime example of this is the Vulpkanin "ear fluff".

**Comment by KasparoVy** [comment]
To open up horns to other species all I'd have to do is add the other species to species_allowed in sprite_accessories and to the species checks in preferences and such, I tried to generalize it as much as possible by working it in with a flag called 'HAS_HORNS' too with just that in mind-- I made the sprite names pretty universal, too

Or are you referring to the frills instead?

**Comment by Fox-McCloud** [comment]
Hmmm; mind testing what the striping looks like on Vulps and Tajarans? That may be something they would both like, provided it fits well.

**Comment by KasparoVy** [comment]
Sure, I'll see what I can do in that regard. I'll finish work on the SQL bit first.

I'm sure tiger striping would look great on Tajarans, not so sure it'd make sense on Vulpkanin.

I'll post some pics of what it all looks like when I'm through

**Comment by KasparoVy** [comment]
@Fox-McCloud Done! Sorry it took so long to push through, took me quite a bit of time to figure out how to get horns and markings to reset properly when species was changed in preferences (to prevent Vulpkanin from having Unathi horns, and Unathi from having Tajara markings etc. etc.)

Here are the photos, as promised. Every new feature is exhibited, head accessories (incl. earfluff for Vulpkanin + Tajara) and markings (tweaked to fit the unique heads of each species).

*Character Creation*
headacccustomizationpoc1

*ID Generation*
headacccustomizationpoc2

*Unathi Ingame*
headacccustomizationpoc5

*Vulpkanin Ingame*
headacccustomizationpoc4

*Tajara Ingame -- (Markings look darker because I messed up the colours in preferences by accident. The base sprites are the same colour.)*
headacccustomizationpoc3

The designs of the characters were made to show what the same colours and most similar features would look like on each of the species

**Comment by KasparoVy** [comment]
Probably going to have a merge conflict with Aurorablade's #2903. Going to sort that out now

**Comment by Fox-McCloud** [comment]
That's pretty great--I think that should definitely be a thing for both Species.

**Comment by KasparoVy** [comment]
I'm glad it turned out this way haha, means that if anyone wanted to create some extra body markings for any of these species the framework's all there and all it takes is some sprites and basic code

Didn't foresee this end result when I first started work on this but I'm quite pleased

**Comment by Crazylemon64** [comment]
Mmm, let me see, where we might conflict...

**Comment by Fox-McCloud** [comment]
>I'm glad it turned out this way, means that if anyone wanted to create some extra body markings for any of these species the framework's all there and all it takes is some sprites and basic code

Yep; it's pretty great--could do things like belly markings for Tajaran/Vulps/Unathi, splotches, etc.

One question though; why the inclusion of Tajarans Ears and Vulp earfluff in both facial hair and head accessory?

**Comment by KasparoVy** [comment]
I left that in as afterthought in case someone preferred it the way it was before this PR, but I could remove them from facial hair if you'd like-- I just wasn't certain about that bit there so I felt this'd be a safe way to go about it

**Comment by KasparoVy** [comment]
@Crazylemon64 I'm going to take a look at that file now, never had to mess around with SQL stuff before myself

**Comment by KasparoVy** [comment]
Killed the old earfluff entries in the Tajaran hair and Vulpkanin facial hair sections, finished SQL support as per @Crazylemon64's pointer (thanks! I totally missed that) and attempted to accommodate his PR so if this gets merged I won't break his stuff. Not sure if I missed anything in that regard

**Comment by Crazylemon64** [comment]
Ah, that looks good! Fantastic! Though, the merge won't go smoothly until the sql_preferences.dm file is fully merged; I think you can go ahead first, and I can make the tweaks needed to my edition to get things going smoothly.

**Comment by KasparoVy** [comment]
Oh, I was about to push a commit that modified preferences_mysql.dm to accommodate your changes. Was that file not its final version?

**Comment by Crazylemon64** [comment]
Oh, if it does that, then we should be golden. Go for it!

**Comment by KasparoVy** [comment]
Alrighty! This will no longer compile, but the plan is-- once your PR goes through, all I'll need to do is update my branch from master and I'll inherit the rest of what you've done and THEN it will compile and be all good.

**Comment by KasparoVy** [comment]
Sweet, done. This will no longer compile, but the plan is-- once your PR goes through, all I'll need to do is update my branch from master and I'll inherit the rest of what you've done and THEN it will compile and be all good. So we're set. Thanks, I'm glad we could work this out!

**This should definitely not be merged until #3025 has been and this PR has seen one last commit to update**

**Comment by KasparoVy** [comment]
Sweet, done. This will no longer compile, but the plan is-- once your PR goes through, all I'll need to do is update my branch from master and I'll inherit the rest of what you've done and THEN it will compile and be all good. So we're set. Thanks, I'm glad we could work this out!

**This should definitely not be merged until #3025 has been and this PR has seen one last commit to update**

**Comment by Fox-McCloud** [comment]
Hm. How is this going to work out for those who have Vulp earfluff and Tajaran "ears" as their facial/hair style now that it's gotten rid of?

Also, one thing worth testing is making sure that some of these icons carry over when you get cloned.

**Comment by KasparoVy** [comment]
Made a new branch and reverted like 4 commits so I could test just that, I'll append the results of that test in just a moment

**Comment by KasparoVy** [comment]
While testing, it appears as though body accessories (tails) are not carried through cloning either.

I believe this to be the case because stuff like that isn't included in UI+UE generation in dna2
So that means if a Vulpkanin with a stubby tail is cloned, they'll come out with a big bushy standard tail.

I'll see if I can fix that, too, but doing so might bring a new host of problems-- one of which being, what if the brain was transplanted into a human body and cloned? Since the body is generated by UI+UE, it'd give the human body a Vulpkanin tail or whatever.

Hm.

**Comment by KasparoVy** [comment]
Tested the most recent commit on my test branch and it worked perfectly, cloned players with markings and/or head accessories will get them back with their new bodies.

This still won't compile properly as I'm waiting on the other PR to go through so I can make one last commit to inherit the files this PR is dependent on

cloningpoc1

Yay!

**Comment by Fox-McCloud** [comment]
Excelllllent.

**Comment by Fox-McCloud** [comment]
mere conflicts.

May just want to start over for this PR.

**Comment by Crazylemon64** [comment]
Mm... which files fail to automerge, when you do "git pull"?

**Comment by KasparoVy** [comment]
Sorry, just working through it all now. I've just merged all the stuff that came through earlier today and I'm seeing if it compiles before pushing

There were two conflicts, /client/preferences.dm and /client/preferences_mysql.dm-- VERY small conflicts.

Everything seems to have compiled nicely, so I'm going to test it one last time and push then we should be okay

**Comment by KasparoVy** [comment]
Nice, I guess that's it then.

**Comment by tigercat2000** [comment]
Okay, I have to be honest with this- This touches a **lot** of stuff for aesthetic features, and on top of that, has 26 commits for 27 files, and that makes me a bit nervous on supporting it fully.

**Comment by KasparoVy** [comment]
It touches a lot of files because, in order to have facial hair overlay head accessories and hair to be overlayed by head accessories, I had to split update_hair into update_hair and update_fhair-- and put facial hair on its own layer.

That means that I had to go through a bunch of files that called update_hair with the intention of updating facial hair and add update_fhair to them

Additionally, the number of commits is a result of continuous tweaking, things I'd missed, and maintenance updates in order to keep up with other PRs

A couple PRs were waste because I ended up gutting the extra Unathi tails for a future PR anyway, so thats' where a few of them came from

Otherwise the end result is what's in the pictures, head accessories, species flags for those accessories, Unathi-specific facial hair, a refactor of one Vulpkanin and Tajaran facial hair style each, SQL and cloning support for those things

**Comment by tigercat2000** [comment]
@KasparoVy I have no doubt that it does _work_ just fine, but it worries me with some things, like touching DNA code- something I'm aware has to be done for changes to persist through cloning, however, DNA code is a nightmare on it's own, and that's assuming that it's modular features (IE, new blocks) actually work correctly at all (which isn't guaranteed).

**Comment by KasparoVy** [comment]
All I did with the DNA code was add in the new blocks for the RGB of the head accessories + markings and the style of each, and then worked that in with the update_appearance proc in the helper file

It worked in practice with cloning

After cloning I checked each individual variable on my character (the cloned body) and the r/g/b_headacc and r/g/b_markings had been inherited from the old body's UI, as well as m_style and ha_style

**Comment by Fox-McCloud** [comment]
@KasparoVy 

Any time you're specifically available for testing? I'd like to demo this, myself.

**Comment by KasparoVy** [comment]
Sorry, just woke up-- I'm available from now (5:20pm EST) until about 7 hours from now for testing

**Comment by KasparoVy** [comment]
@Fox-McCloud Reposting this comment, -- I'm available from 5:00pm EST until 3:00am for testing, and it's 9:30pm right now-- both today and tomorrow in the same time range

I'll be checking back periodically, I'm ready whenever

**Comment by Fox-McCloud** [comment]
Tested this out; it works quite well--up to and including the genetics.

May want to double-check the SQL stuff though, @tigercat2000

**Comment by Fox-McCloud** [comment]
@KasparoVy 

Were you going to do that one last thing we discussed? If so, let me know.

**Comment by KasparoVy** [comment]
@Fox-McCloud I'm more than happy to do it, sent you a forum PM with the results of the testing enclosed. Just for posterity, here's another copy
facemarkingspoc1
1. Head accessory: Kita, Facial hair: Blaze
2. Head accessory: Mask, Facial hair: Patch
3. Head accessory: Vulpine, Facial hair: Swift
4. Head accessory: Vulpine, Facial hair: Patch
5. Head accessory: Mask, Facial hair: Kita

**Pull request merged**